examples/games/mafia_werewolf/play_mafia.qnt

```diff
@@ -0,0 +1,185 @@
+                val votes = votes_by_player.get(p)
+                if (votes > acc) votes else acc
+            })
+            PLAYERS.filter(p => votes_by_player.get(p) == max_votes)
+        } else Set()  // Return an empty set if not all players have voted
+    }
+    /// Check if all Mafia players are dead
+    pure def all_mafias_dead(player: str -> PlayerFeatures): bool = {
+        PLAYERS.filter(p => player.get(p).role == Mafia).forall(p => player.get(p).status == Dead)
+    }
+    /// Check if all Citizen players are dead
+    pure def all_citizens_dead(player: str -> PlayerFeatures): bool = {
+        PLAYERS.filter(p => player.get(p).role == Citizen).forall(p => player.get(p).status == Dead)
+    }
+    /// Update the game status based on the current state
+    pure def update_status(player: str -> PlayerFeatures): Status = {
+        if (all_mafias_dead(player)) Done(Citizen)  // Citizens win if all Mafias are dead
+        else if (all_citizens_dead(player)) Done(Mafia)  // Mafia wins if all Citizens are dead
+        else Pending  // The game is still ongoing
+    }
+    /// Function to reset the players votes after voting is done
+    pure def reset_votes_and_statuses(player_to_features: str -> PlayerFeatures): str -> PlayerFeatures = {
+        PLAYERS.mapBy(p => {
+            ...player_to_features.get(p),
+            voted: false
+        })
+    }
+    /// Function to update player features after hanging
+    def update_features_after_hang(player_to_hang: str): str -> PlayerFeatures = {
+        PLAYERS.mapBy(p =>
+            if (p == player_to_hang) {
+                ...player_to_features.get(p),
+                status: Dead,
+                voted: false
+            } else {
+                ...player_to_features.get(p),
+                voted: false
+            }
+        )
+    }
+    /// Function to update player features after being killed by the mafia
+    def update_features_after_kill(victim: str): str -> PlayerFeatures ={
+        player_to_features.set(victim, {
+            ...player_to_features.get(victim),
+            status: Dead
+        })
+    }
+
+    action init = all {
+        pure val roles = Set(Mafia, Citizen)
+        nondet role_by_player = PLAYERS.setOfMaps(roles).oneOf()
+        player_to_features' = PLAYERS.mapBy(name => { role: role_by_player.get(name), status: Alive, voted: false }),
+        game_phase' = Night,  // Start with the Night phase
+        game_status' = Pending,  // Game is in Pending status
+        votes_by_player' = PLAYERS.mapBy(p => 0)  // Initialize vote counts to 0
+    }
+
+    /// Action for Mafia killing a player
+    action mafia_kills = any {
+        nondet killer = PLAYERS.filter(p => player_to_features.get(p).status == Alive and player_to_features.get(p).role == Mafia).oneOf()
+        nondet victim = PLAYERS.filter(p => player_to_features.get(p).status == Alive and player_to_features.get(p).role == Citizen).oneOf()
```

Comment:
It might be easier for the whole spec in general if we just use and pass around the actual `PlayerFeatures` value instead of the player name, so you don't need to be calling `get` all the time - which will result in a clearer and faster spec. With the change below, `killer` and `victim` will be of type `PlayerFeatures`. The only problem is that you lose the `name` information, which you need in order to update the `player_to_features` map in the right key. What I would do is to add `name: str` to `PlayerFeatures`, and then we can have everything in one value.

We can talk more about this in our meeting tomorrow

```suggestion
        nondet killer = player_to_features.values().filter(p => p.status == Alive and p.role == Mafia).oneOf()
        nondet victim = player_to_features.filter(p => pp.status == Alive and p.role == Citizen).oneOf()
```

and then, in `update_features_after_kill`, you'd have something like:
```
player_to_features.set(victim.name, { ...victim, status: Dead })
```